server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java

```diff
@@ -73,26 +86,56 @@ protected RowRecord nextWithoutConstraint() throws IOException {
    * @return if there has next row record.
    */
   private boolean cacheRowRecord() throws IOException {
-    while (timeGenerator.hasNext()) {
-      boolean hasField = false;
-      long timestamp = timeGenerator.next();
-      RowRecord rowRecord = new RowRecord(timestamp);
-      for (int i = 0; i < seriesReaderByTimestampList.size(); i++) {
-        IReaderByTimestamp reader = seriesReaderByTimestampList.get(i);
-        Object value = reader.getValueInTimestamp(timestamp);
-        if (value == null) {
-          rowRecord.addField(null);
-        } else {
-          hasField = true;
-          rowRecord.addField(value, dataTypes.get(i));
+    int seriesNum = seriesReaderByTimestampList.size();
+    while (timeGenerator.hasNextTimeColumn()) {
+      final TimeColumn timeColumn = timeGenerator.nextTimeColumn();
+
+      Future<List<Field>>[] futures = new Future[seriesNum];
+      for (int i = 0; i < seriesNum; i++) {
+        final IReaderByTimestamp readerByTimestamp = seriesReaderByTimestampList.get(i);
+        final TSDataType tsDataType = dataTypes.get(i);
+        futures[i] = QueryTaskPoolManager.getInstance().submit(() -> {
+          List<Field> fields = new ArrayList<>();
+          Object[] values = readerByTimestamp
+              .getValuesInTimestamps(timeColumn.duplicate(), Long.MAX_VALUE);
+          for (Object value : values) {
+            if (value == null) {
+              fields.add(null);
+            } else {
+              fields.add(Field.getField(value, tsDataType));
+            }
+          }
+          values = null;
```

Comment:
This `values` is not used, maybe it could be removed?

Reply:
I think this is in order to reduce the pressure of GC, but it looks as if there is no escape. I've deleted it